perf/src/qos.rs

```diff
@@ -0,0 +1,121 @@
+use {
+    crate::{
+        data_budget::DataBudget,
+        packet::{Packet, PacketBatch},
+    },
+    std::{
+        collections::HashMap,
+        net::IpAddr,
+        sync::atomic::{AtomicUsize, Ordering},
+    },
+};
+
+pub type StakeMap = HashMap<IpAddr, (u64, DataBudget)>;
+
+fn qos_packet(
+    num_discarded: &AtomicUsize,
+    max_packets_per_second: u64,
+    default_budget: &DataBudget,
+    total_stake: u64,
+    ip_to_budget: &StakeMap,
+    packet: &mut Packet,
+) {
+    if packet.meta.discard() {
+        return;
+    }
+
+    if let Some((stake, budget)) = ip_to_budget.get(&packet.meta.addr) {
```

Comment:
@sakridge can we run this pre dedup and punish the IPs that are flooding the network?